File: frontend/src/utils/validation.ts
export interface ValidationResult {
  isValid: boolean;
  errors: string[];
}

export const validateUsername = (username: string): ValidationResult => {
  const errors: string[] = [];

  if (!username || username.trim() === '') {
    errors.push('Username is required');
  } else {
    if (username.length < 3) {
      errors.push('Username must be at least 3 characters');
    }
    if (username.length > 50) {
      errors.push('Username must be at most 50 characters');
    }
    if (!/^[a-zA-Z0-9_-]+$/.test(username)) {
      errors.push('Username can only contain letters, numbers, underscores, and hyphens');
    }
  }

  return { isValid: errors.length === 0, errors };
};

export const validateEmail = (email: string): ValidationResult => {
  const errors: string[] = [];

  if (!email || email.trim() === '') {
    errors.push('Email is required');
  } else {
    const emailRegex = /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/;
    if (!emailRegex.test(email)) {
      errors.push('Invalid email format');
    }
  }

  return { isValid: errors.length === 0, errors };
};

export const validatePassword = (password: string): ValidationResult => {
  const errors: string[] = [];

  if (!password || password.trim() === '') {
    errors.push('Password is required');
  } else {
    if (password.length < 8) {
      errors.push('Password must be at least 8 characters long');
    }
    if (!/[A-Z]/.test(password)) {
      errors.push('Password must contain at least one uppercase letter');
    }
    if (!/[a-z]/.test(password)) {
      errors.push('Password must contain at least one lowercase letter');
    }
    if (!/[0-9]/.test(password)) {
      errors.push('Password must contain at least one number');
    }
    if (!/[!@#$%^&*()_+\-=\[\]{};':"\\|,.<>\/?~`]/.test(password)) {
      errors.push('Password must contain at least one special character');
    }
  }

  return { isValid: errors.length === 0, errors };
};

export const validateName = (name: string, fieldName: string): ValidationResult => {
  const errors: string[] = [];

  if (!name || name.trim() === '') {
    errors.push(`${fieldName} is required`);
  } else {
    if (name.length < 2) {
      errors.push(`${fieldName} must be at least 2 characters`);
    }
    if (name.length > 100) {
      errors.push(`${fieldName} must be at most 100 characters`);
    }
  }

  return { isValid: errors.length === 0, errors };
};

export const getPasswordStrength = (password: string): { strength: number; label: string; color: string } => {
  let strength = 0;

  if (password.length >= 8) strength++;
  if (password.length >= 12) strength++;
  if (/[A-Z]/.test(password)) strength++;
  if (/[a-z]/.test(password)) strength++;
  if (/[0-9]/.test(password)) strength++;
  if (/[!@#$%^&*()_+\-=\[\]{};':"\\|,.<>\/?~`]/.test(password)) strength++;

  if (strength <= 2) return { strength, label: 'Weak', color: '#dc3545' };
  if (strength <= 4) return { strength, label: 'Medium', color: '#ffc107' };
  return { strength, label: 'Strong', color: '#28a745' };
};

File: backend/internal/services/validation.go
package services

import (
	"errors"
	"regexp"
	"strings"

	"github.com/google/uuid"
	"github.com/seasia/auth-service/internal/models"
	"gorm.io/gorm"
)

type ValidationService struct {
	db *gorm.DB
}

func NewValidationService(db *gorm.DB) *ValidationService {
	return &ValidationService{db: db}
}

type ValidationError struct {
	Field   string `json:"field"`
	Message string `json:"message"`
}

func (v *ValidationService) ValidateUsername(username string, excludeUserID *uuid.UUID) error {
	if username == "" {
		return errors.New("username is required")
	}
	if len(username) < 3 {
		return errors.New("username must be at least 3 characters")
	}
	if len(username) > 50 {
		return errors.New("username must be at most 50 characters")
	}
	if !regexp.MustCompile(`^[a-zA-Z0-9_-]+$`).MatchString(username) {
		return errors.New("username can only contain letters, numbers, underscores, and hyphens")
	}

	var count int64
	query := v.db.Model(&models.User{}).Where("LOWER(username) = LOWER(?)", username)
	if excludeUserID != nil {
		query = query.Where("id != ?", *excludeUserID)
	}
	query.Count(&count)

	if count > 0 {
		return errors.New("username already exists")
	}
	return nil
}

func (v *ValidationService) ValidateEmail(email string, excludeUserID *uuid.UUID) error {
	if email == "" {
		return errors.New("email is required")
	}
	email = strings.ToLower(strings.TrimSpace(email))
	if !regexp.MustCompile(`^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$`).MatchString(email) {
		return errors.New("invalid email format")
	}

	var count int64
	query := v.db.Model(&models.User{}).Where("LOWER(email) = LOWER(?)", email)
	if excludeUserID != nil {
		query = query.Where("id != ?", *excludeUserID)
	}
	query.Count(&count)

	if count > 0 {
		return errors.New("email already exists")
	}
	return nil
}

func (v *ValidationService) ValidateRoleName(name string, excludeRoleID *uuid.UUID) error {
	if name == "" {
		return errors.New("role name is required")
	}
	if len(name) < 2 {
		return errors.New("role name must be at least 2 characters")
	}
	if len(name) > 100 {
		return errors.New("role name must be at most 100 characters")
	}

	var count int64
	query := v.db.Model(&models.Role{}).Where("LOWER(name) = LOWER(?)", name)
	if excludeRoleID != nil {
		query = query.Where("id != ?", *excludeRoleID)
	}
	query.Count(&count)

	if count > 0 {
		return errors.New("role name already exists")
	}
	return nil
}

func (v *ValidationService) ValidateGroupName(name string, excludeGroupID *uuid.UUID) error {
	if name == "" {
		return errors.New("group name is required")
	}
	if len(name) < 2 {
		return errors.New("group name must be at least 2 characters")
	}
	if len(name) > 100 {
		return errors.New("group name must be at most 100 characters")
	}

	var count int64
	query := v.db.Model(&models.Group{}).Where("LOWER(name) = LOWER(?)", name)
	if excludeGroupID != nil {
		query = query.Where("id != ?", *excludeGroupID)
	}
	query.Count(&count)

	if count > 0 {
		return errors.New("group name already exists")
	}
	return nil
}

func (v *ValidationService) ValidateServiceName(name string, excludeServiceID *uuid.UUID) error {
	if name == "" {
		return errors.New("service name is required")
	}
	if len(name) < 2 {
		return errors.New("service name must be at least 2 characters")
	}
	if len(name) > 100 {
		return errors.New("service name must be at most 100 characters")
	}

	var count int64
	query := v.db.Model(&models.Service{}).Where("LOWER(name) = LOWER(?)", name)
	if excludeServiceID != nil {
		query = query.Where("id != ?", *excludeServiceID)
	}
	query.Count(&count)

	if count > 0 {
		return errors.New("service name already exists")
	}
	return nil
}

type PasswordValidation struct {
	MinLength                int
	RequireUppercase         bool
	RequireLowercase         bool
	RequireNumbers           bool
	RequireSpecialCharacters bool
}

func (v *ValidationService) ValidatePassword(password string, config PasswordValidation) []string {
	var errors []string

	if len(password) < config.MinLength {
		errors = append(errors, "password must be at least 8 characters long")
	}

	if config.RequireUppercase && !regexp.MustCompile(`[A-Z]`).MatchString(password) {
		errors = append(errors, "password must contain at least one uppercase letter")
	}

	if config.RequireLowercase && !regexp.MustCompile(`[a-z]`).MatchString(password) {
		errors = append(errors, "password must contain at least one lowercase letter")
	}

	if config.RequireNumbers && !regexp.MustCompile(`[0-9]`).MatchString(password) {
		errors = append(errors, "password must contain at least one number")
	}

	if config.RequireSpecialCharacters && !regexp.MustCompile(`[!@#$%^&*()_+\-=\[\]{};':"\\|,.<>\/?~` + "`" + `]`).MatchString(password) {
		errors = append(errors, "password must contain at least one special character")
	}

	return errors
}
